internal/app/uploader.go

```diff
@@ -0,0 +1,85 @@
+package app
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"golang.org/x/crypto/sha3"
+)
+
+// BasinProviderUploader ...
+type BasinProviderUploader interface {
+	Upload(context.Context, string, string, io.Reader, *Signer, io.Writer) error
+	Reconnect() error
+}
+
+// BasinUploader contains logic of uploading Parquet files to Basin Provider.
+type BasinUploader struct {
+	namespace  string
+	relation   string
+	privateKey *ecdsa.PrivateKey
+	provider   BasinProviderUploader
+}
+
+// NewBasinUploader creates new uploader.
+func NewBasinUploader(ns string, rel string, bp BasinProviderUploader, pk *ecdsa.PrivateKey) *BasinUploader {
+	return &BasinUploader{
+		namespace:  ns,
+		relation:   rel,
+		provider:   bp,
+		privateKey: pk,
+	}
+}
+
+// Upload sends file to provider for upload.
+func (bu *BasinUploader) Upload(ctx context.Context, filepath string, progress io.Writer) error {
+	f, err := os.Open(filepath)
+	if err != nil {
+		return fmt.Errorf("open file: %s", err)
+	}
+	defer func() {
+		_ = f.Close()
+	}()
+
+	if err := bu.provider.Upload(ctx, bu.namespace, bu.relation, f, NewSigner(bu.privateKey), progress); err != nil {
+		return fmt.Errorf("upload: %s", err)
+	}
+
+	return nil
+}
+
+// Signer allows you to sign a big stream of bytes by calling Sum multiple times, then Sign.
+type Signer struct {
```

Comment:
it is an implementation of an incremental signer that is passed to `basinprovider.Upload` method (L49)